### ISSUE TITLE
grouped_gemm_nt_masked NaN in scale propagates across experts, also affects finite values depending on problem size

### ISSUE THREAD

**Issue by elvircrn** [flashinfer-ai/flashinfer#3103]
In the following repro script, both experts get identical fp4 data and weights while only expert 0's scale is changed. Expert 1's output should never change, and yet a NaN is present in both expert outputs. Furthermore, even with a finite value set for scale, the output is also affect over multiple experts.

```
"""
    python3 repro_gemm_scale_leak.py
"""
import torch, os
os.environ["CUDA_VISIBLE_DEVICES"] = "0"
from flashinfer.cute_dsl.blockscaled_gemm import grouped_gemm_nt_masked

dev = torch.device("cuda:0")
m, n, k, num_experts, sf_vec_size = 128, 128, 256, 2, 16
k_packed = k // 2
rm = m // 128
k_sf = k // sf_vec_size
rk = k_sf // 4
sf_shape = (32, 4, rm, 4, rk, num_experts)

# Same data for both experts
aq = torch.full((m, k_packed, num_experts), 0x11, dtype=torch.uint8, device=dev)
w  = torch.full((n, k_packed, num_experts), 0x11, dtype=torch.uint8, device=dev)
w_bs = torch.ones(num_experts, n, k_sf, dtype=torch.float8_e4m3fn, device=dev)
masked_m = torch.full((num_experts,), m, dtype=torch.int32, device=dev)
alpha = torch.ones(1, 1, num_experts, dtype=torch.float32, device=dev)

def run(scale_byte):
    sf = torch.ones(sf_shape, dtype=torch.float8_e4m3fn, device=dev)
    raw = sf.view(torch.uint8)
    raw[0, 0, 0, 0, 0, 0] = scale_byte  # expert 0 only
    sf = raw.view(torch.float8_e4m3fn).reshape(sf_shape)
    out = torch.zeros(m, n, num_experts, dtype=torch.bfloat16, device=dev)
    grouped_gemm_nt_masked(
        (aq, sf), (w, w_bs), out, masked_m,
        ab_dtype="float4_e2m1fn", sf_dtype="float8_e4m3fn",
        c_dtype="bfloat16", sf_vec_size=sf_vec_size,
        alpha=alpha, alpha_dtype="float32",
    )
    torch.cuda.synchronize()
    return out.permute(2, 0, 1)  # [experts, m, n]

baseline = run(0x3C)  # all scales = 1.0
modified = run(0x40)  # expert 0 scale = 2.0
nan_run  = run(0x7F)  # expert 0 scale = NaN

# Cross-run: expert 1's scale is always 1.0, so its output should never change
base_bits = baseline[1].view(torch.uint16)
print("Expert 1 across runs (should be 0 if no leak):")
for name, out in [("scale=2.0", modified), ("scale=NaN", nan_run)]:
    bits = out[1].view(torch.uint16)
    diff = bits != base_bits
    print(f"  vs baseline: {name:>10} -> {diff.any(dim=-1).sum().item()} rows changed")

# Within-run: if leak copies same value, expert 0 == expert 1 (should differ if no leak)
print("\nExpert 0 vs expert 1 within run (should differ if scales differ, 0 = leak):")
for name, out in [("scale=1.0", baseline), ("scale=2.0", modified), ("scale=NaN", nan_run)]:
    e0_bits = out[0].view(torch.uint16)
    e1_bits = out[1].view(torch.uint16)
    diff = (e0_bits != e1_bits).any(dim=-1).sum().item()
    print(f"  {name:>10} -> {diff} rows differ")
```

Increasing the problem dimensions limits the area of effect, but it seems the kernel is still producing erroneous outputs:

```
"""
Does changing expert 0's scale affect expert 1's output?

Both experts get identical fp4 data and weights.
Only expert 0's scale is changed. Expert 1's output should never change.

    python3 repro_gemm_scale_leak.py
"""
import torch, os
os.environ["CUDA_VISIBLE_DEVICES"] = "0"
from flashinfer.cute_dsl.blockscaled_gemm import grouped_gemm_nt_masked

dev = torch.device("cuda:0")
m, n, k, num_experts, sf_vec_size = 8192, 4096, 7168, 64, 16
k_packed = k // 2
rm = m // 128
k_sf = k // sf_vec_size
rk = k_sf // 4
sf_shape = (32, 4, rm, 4, rk, num_experts)

# Same data for both experts
aq = torch.full((m, k_packed, num_experts), 0x11, dtype=torch.uint8, device=dev)
w  = torch.full((n, k_packed, num_experts), 0x11, dtype=torch.uint8, device=dev)
w_bs = torch.ones(num_experts, n, k_sf, dtype=torch.float8_e4m3fn, device=dev)
masked_m = torch.full((num_experts,), m, dtype=torch.int32, device=dev)
alpha = torch.ones(1, 1, num_experts, dtype=torch.float32, device=dev)

def run(scale_byte):
    sf = torch.ones(sf_shape, dtype=torch.float8_e4m3fn, device=dev)
    raw = sf.view(torch.uint8)
    raw[0, 0, 0, 0, 0, 0] = scale_byte  # expert 0 only
    sf = raw.view(torch.float8_e4m3fn).reshape(sf_shape)
    out = torch.zeros(m, n, num_experts, dtype=torch.bfloat16, device=dev)
    grouped_gemm_nt_masked(
        (aq, sf), (w, w_bs), out, masked_m,
        ab_dtype="float4_e2m1fn", sf_dtype="float8_e4m3fn",
        c_dtype="bfloat16", sf_vec_size=sf_vec_size,
        alpha=alpha, alpha_dtype="float32",
    )
    torch.cuda.synchronize()
    return out.permute(2, 0, 1)  # [experts, m, n]

baseline = run(0x3C)  # all scales = 1.0
modified = run(0x40)  # expert 0 scale = 2.0
nan_run  = run(0x7F)  # expert 0 scale = NaN

# Cross-run: experts 1-7 scales are always 1.0, their output should never change
print("Experts 1-7 across runs (should be 0 if no leak):")
for name, out in [("scale=2.0", modified), ("scale=NaN", nan_run)]:
    leaked = []
    for e in range(1, num_experts):
        diff = baseline[e].view(torch.uint16) != out[e].view(torch.uint16)
        rows = diff.any(dim=-1).sum().item()
        if rows > 0:
            leaked.append(f"e{e}={rows}")
    print(f"  vs baseline: {name:>10} -> {', '.join(leaked) if leaked else 'no leak'}")

# Within-run: expert 0 vs each other expert (should differ if scales differ, 0 = leak)
print("\nExpert 0 vs others within run (0 rows differ = leak):")
for name, out in [("scale=1.0", baseline), ("scale=2.0", modified), ("scale=NaN", nan_run)]:
    diffs = []
    for e in range(1, num_experts):
        d = (out[0].view(torch.uint16) != out[e].view(torch.uint16)).any(dim=-1).sum().item()
        diffs.append(f"e{e}={d}")
    print(f"  {name:>10} -> {', '.join(diffs)}")
```

**Comment by kahyunnam** [comment]
This seems to be an output-layout contract bug (assumed but missing verification check in code), not a padding/NaN issue. `grouped_gemm_nt_masked` hard-codes c_major="n" and builds the C tensor from the output pointer with an assumed layout (order=(1,0,2)), ignoring the output's actual strides. The C store is TMA over (M,N) tiles, so the output's contiguous dim must be N. 

Meanwhile, the repro allocates `out = torch.zeros(m, n, num_experts, ...etc)`  whcih is expert dim innermost, so each expert's tile is stored to interleaved offsets and clobbers the others.
```
# leaks:
out = torch.zeros(m, n, num_experts, dtype=torch.bfloat16, device=dev)
# clean (N-contiguous, expert dim outermost):
out = torch.zeros(num_experts, m, n, dtype=torch.bfloat16, device=dev).permute(1, 2, 0)
```

Fix: grouped_gemm_nt_masked now validates the output layout and raises instead of corrupting; callers must pass an N-contiguous output (e.g. torch.empty(l, m, n).permute(1, 2, 0)). Added a regression test.

**Comment by kahyunnam** [comment]
Repro script to verify bug against fix PR #3574 : 
```
"""
Verify the #3103 fix against the issue's ORIGINAL repro.

Part 1: original output layout `out = torch.zeros(m, n, num_experts)` (expert
        dim innermost). With the fix, grouped_gemm_nt_masked now REJECTS it.
Part 2: same repro, output allocation corrected to be N-contiguous /
        expert-outermost; perturbing expert 0 must not affect expert 1.

e4m3 scale bytes: 0x38 = 1.0, 0x40 = 2.0, 0x7F = NaN.
(The original gist used 0x3C as "1.0"; 0x3C is actually 1.5, which is why its
within-run baseline showed a nonzero diff. We use 0x38 for a true no-op baseline.)

    cd /workspace/flashinfer
    python repro_issue_3103_verify.py
"""
import torch
from flashinfer.cute_dsl.blockscaled_gemm import grouped_gemm_nt_masked

dev = torch.device("cuda:0")
m, n, k, num_experts, sf_vec_size = 128, 128, 256, 2, 16
k_packed = k // 2
rm = m // 128
k_sf = k // sf_vec_size
rk = k_sf // 4
sf_shape = (32, 4, rm, 4, rk, num_experts)

aq = torch.full((m, k_packed, num_experts), 0x11, dtype=torch.uint8, device=dev)
w = torch.full((n, k_packed, num_experts), 0x11, dtype=torch.uint8, device=dev)
w_bs = torch.ones(num_experts, n, k_sf, dtype=torch.float8_e4m3fn, device=dev)
masked_m = torch.full((num_experts,), m, dtype=torch.int32, device=dev)
alpha = torch.ones(1, 1, num_experts, dtype=torch.float32, device=dev)


def run(scale_byte, make_out):
    sf = torch.ones(sf_shape, dtype=torch.float8_e4m3fn, device=dev)
    raw = sf.view(torch.uint8)
    raw[0, 0, 0, 0, 0, 0] = scale_byte  # perturb expert 0's first block only
    sf = raw.view(torch.float8_e4m3fn).reshape(sf_shape)
    out = make_out()
    grouped_gemm_nt_masked(
        (aq, sf), (w, w_bs), out, masked_m,
        ab_dtype="float4_e2m1fn", sf_dtype="float8_e4m3fn",
        c_dtype="bfloat16", sf_vec_size=sf_vec_size,
        alpha=alpha, alpha_dtype="float32",
    )
    torch.cuda.synchronize()
    return out.permute(2, 0, 1)  # [experts, m, n]


def leak_checks(make_out):
    baseline = run(0x38, make_out)  # 1.0 (no-op: all scales stay 1.0)
    modified = run(0x40, make_out)  # expert 0 scale = 2.0
    nan_run = run(0x7F, make_out)   # expert 0 scale = NaN

    # Cross-run: expert 1's scale never changes, so its output must be identical
    # across runs. Any change == cross-expert leak.
    base_e1 = baseline[1].view(torch.uint16)
    print("Expert 1 across runs (must be 0 -- it never changed):")
    leaked = False
    for name, out in [("scale=2.0", modified), ("scale=NaN", nan_run)]:
        rows = (out[1].view(torch.uint16) != base_e1).any(dim=-1).sum().item()
        leaked |= rows > 0
        print(f"  vs baseline: {name:>9} -> {rows} rows changed")

    # Within-run: expert 0 (perturbed) vs expert 1 (clean). Should differ ONLY at
    # the single perturbed row when the scale differs; 1.0 baseline is identical.
    print("Expert 0 vs expert 1 within run (expected: 0 / 1 / 1):")
    for name, out in [("scale=1.0", baseline), ("scale=2.0", modified), ("scale=NaN", nan_run)]:
        d = (out[0].view(torch.uint16) != out[1].view(torch.uint16)).any(dim=-1).sum().item()
        print(f"  {name:>9} -> {d} rows differ")
    return not leaked


print("=" * 70)
print("PART 1: original layout  out = torch.zeros(m, n, num_experts)")
print("=" * 70)
try:
    leak_checks(lambda: torch.zeros(m, n, num_experts, dtype=torch.bfloat16, device=dev))
    print(">>> FAIL: ran without error (fix not active / stale JIT cache?)")
except ValueError as e:
    print(f">>> REJECTED by the fix (expected):\n    {e}")

print("\n" + "=" * 70)
print("PART 2: corrected layout  out = torch.zeros(num_experts, m, n).permute(1, 2, 0)")
print("=" * 70)
ok = leak_checks(
    lambda: torch.zeros(num_experts, m, n, dtype=torch.bfloat16, device=dev).permute(1, 2, 0)
)
print(f">>> {'PASS: no cross-expert leak' if ok else 'FAIL: expert 1 changed (leak)'}")

```

Output on B200, CUDA 13.0 flashinfer container:
```
======================================================================
PART 1: original layout  out = torch.zeros(m, n, num_experts)
======================================================================
>>> REJECTED by the fix (expected):
    grouped_gemm_nt_masked: `out` must be a 3D (m, n, l) tensor that is contiguous in the layout implied by c_major='n' (the 'n' dimension contiguous and the batch/expert dim outermost). For c_major='n' allocate e.g. `torch.empty(l, m, n, ...).permute(1, 2, 0)`. Got out.shape=(128, 128, 2), out.stride()=(256, 2, 1). A non-compliant layout (e.g. the expert dim innermost) causes silent cross-expert output corruption.

======================================================================
PART 2: corrected layout  out = torch.zeros(num_experts, m, n).permute(1, 2, 0)
======================================================================
Expert 1 across runs (must be 0 -- it never changed):
  vs baseline: scale=2.0 -> 0 rows changed
  vs baseline: scale=NaN -> 0 rows changed
Expert 0 vs expert 1 within run (expected: 0 / 1 / 1):
  scale=1.0 -> 0 rows differ
  scale=2.0 -> 1 rows differ
  scale=NaN -> 1 rows differ
>>> PASS: no cross-expert leak
```

**Comment by elvircrn** [comment]
@kahyunnam Thank you for digging into this!

**Comment by kahyunnam** [comment]
Closing as fixed is merged
